File: src/controllers/JokenpoUsecase.java
package controllers;

import models.ActionResult;
import utils.CircleList;
import utils.PlayChoice;
import utils.PlayResults;

public class JokenpoUsecase {

    final CircleList<PlayChoice> possiblePlays;

    public JokenpoUsecase() {

        possiblePlays = new CircleList<>(PlayChoice.PAPER);
        possiblePlays.add(PlayChoice.ROCK);
        possiblePlays.add(PlayChoice.SCISSOR);
    }

    public PlayChoice fromNumber(int number) {
        final PlayChoice[] choices = PlayChoice.values();
        if (choices.length <= number) {
            return choices[0];
        }
        return choices[number];
    }

    public ActionResult execute(final PlayChoice playerChoice, final PlayChoice computerChoice) {
        while (possiblePlays.getData() != playerChoice)
            possiblePlays.next();

        if (computerChoice == playerChoice) {
            return new ActionResult(
                    PlayResults.DRAW,
                    "Empate, ambos selecionaram " + playerChoice.name()
            );
        } else if (computerChoice == possiblePlays.inNext()) {
            return new ActionResult(
                    PlayResults.LOSE,
                    computerChoice.name() + " ganha de " + playerChoice.name()
            );
        } else {
            return new ActionResult(
                    PlayResults.WIN,
                    playerChoice.name() + " ganha de " + computerChoice.name()
            );
        }
    }
}

File: src/controllers/JokenpoUsecase.kt
package controllers

import models.ActionResult
import utils.CircleList
import utils.PlayChoice
import utils.PlayResults

class JokenpoUsecase {
    private val possiblePlays: CircleList<PlayChoice> = CircleList(PlayChoice.PAPER)

    init {
        possiblePlays.add(PlayChoice.ROCK)
        possiblePlays.add(PlayChoice.SCISSOR)
    }

    fun fromNumber(number: Int): PlayChoice {
        val choices = PlayChoice.values()
        return if (choices.size <= number) {
            choices[0]
        } else choices[number]
    }

    fun execute(playerChoice: PlayChoice, computerChoice: PlayChoice?): ActionResult {
        while (possiblePlays.data !== playerChoice) possiblePlays.next()
        return when (computerChoice) {
            playerChoice -> {
                ActionResult(
                    PlayResults.DRAW,
                    "Empate, ambos selecionaram " + playerChoice.name
                )
            }
            possiblePlays.inNext() -> {
                ActionResult(
                    PlayResults.LOSE,
                    computerChoice!!.name + " ganha de " + playerChoice.name
                )
            }
            else -> {
                ActionResult(
                    PlayResults.WIN,
                    playerChoice.name + " ganha de " + computerChoice!!.name
                )
            }
        }
    }
}
